File: 2023/day12.py
import os
import sys
import functools

test = """\
???.### 1,1,3
.??..??...?##. 1,1,3
?#?#?#?#?#?#?#? 1,3,1,6
????.#...#... 4,1,1
????.######..#####. 1,6,5
?###???????? 3,2,1"""

day = os.path.splitext(os.path.basename(__file__))[0]

if 'test' in sys.argv:
    data = test.splitlines()
else:
    data = open(day+'.txt').read().splitlines()

#data = [list(row) for row in data]

DEBUG = 'debug' in sys.argv

# Parse the input.

def parse(data):
    for row in data:
        maps,nums = row.split()
        nums = [int(x) for x in nums.split(',')]
        yield maps,nums

data = list(parse(data))

# Does chk match the first pat of pat?

def match(pat,chk):
    return all( p in (c,'?') for c,p in zip(chk,pat))

# Each call of this does one chunk of #s.  We generage all possible strings that
# end with "###." for this chunk.  If that prefix matches the current spot in the 
# pattern, we recursively try the next.  It's only the memoizing that allows 
# this to run in finite time.

@functools.cache
def gen(pat, size, nums):
    if not nums:
        return all(c in '.?' for c in pat)

    now = nums[0]
    rest = nums[1:]
    after = sum(rest)+len(rest)

    count = 0

    for before in range(size-after-now+1):
        s = '.' * before + '#' * now + '.'
        if match(pat,s):
            count += gen(pat[len(s):], size-now-before-1, rest)
    
    return count        
    
def count_matches(pat,nums):
    return gen(pat, len(pat), tuple(nums))

def part1(data):
    sumx = 0
    for pat,nums in data:
        sumx += count_matches(pat,nums)
    return sumx

def part2(data):
    sumx = 0
    for pat,nums in data:
        pat = '?'.join([pat]*5)
        nums = nums*5
        sumx += count_matches(pat,nums)
    return sumx

print("Part 1:", part1(data))
print("Part 2:", part2(data))

File: 2023/day12.go
package main

import (
	"fmt"
	"strings"

	"aoc/tools"
	_ "embed"
)

var DEBUG bool = false
var TEST bool = false
var test = `
???.### 1,1,3
.??..??...?##. 1,1,3
?#?#?#?#?#?#?#? 1,3,1,6
????.#...#... 4,1,1
????.######..#####. 1,6,5
?###???????? 3,2,1`[1:]

//go:embed day12.txt
var live string

// Does chk match the first pat of pat?

func match(pat string, chk string) bool {
	for i := range min(len(chk), len(pat)) {
		if pat[i] != chk[i] && pat[i] != '?' {
			return false
		}
	}
	return true
}

// Each call of this does one chunk of #s.  We generage all possible strings that
// end with "###." for this chunk.  If that prefix matches the current spot in the
// pattern, we recursively try the next.  Its only the memoizing that allows
// this to run in finite time.

type Cache struct {
	pat  string
	size int
	nums uint64
}

func hash(nums []int) uint64 {
	var h uint64 = 0
	for _, i := range nums {
		h = h*13 + uint64(i)
	}
	return h
}

var cache map[Cache]int

func gen(pat string, size int, nums []int) int {
	if len(nums) == 0 {
		if strings.Contains(pat, "#") {
			return 0
		} else {
			return 1
		}
	}

	entry := Cache{pat, size, hash(nums)}
	if cache[entry] > 0 {
		return cache[entry] - 1
	}

	now := nums[0]
	rest := nums[1:]
	after := tools.Sum(rest) + len(rest)

	count := 0

	for before := range size - after - now + 1 {
		s := strings.Repeat(".", before) + strings.Repeat("#", now) + "."
		if match(pat, s) {
			n := min(len(pat), len(s))
			count += gen(pat[n:], size-now-before-1, rest)
		}
	}
	cache[entry] = count + 1
	return count
}

func count_matches(pat string, nums []int) int {
	return gen(pat, len(pat), nums)
}

type Context struct {
	pat  string
	nums []int
}

func part1(data []Context) int {
	sum := 0
	for _, ctx := range data {
		sum += count_matches(ctx.pat, ctx.nums)
	}
	return sum
}

func part2(data []Context) int {
	sum := 0
	for _, ctx := range data {
		pat := ctx.pat + "?" + ctx.pat + "?" + ctx.pat + "?" + ctx.pat + "?" + ctx.pat
		nums := []int{}
		for range 5 {
			nums = append(nums, ctx.nums...)
		}
		sum += count_matches(pat, nums)
	}
	return sum
}

func main() {
	var input string
	TEST, DEBUG, input = tools.Setup(test, live)

	// Parse the input.

	data := []Context{}
	for _, row := range strings.Split(input, "\n") {
		maps, snums, _ := strings.Cut(row, " ")
		nums := tools.SplitIntBy(snums, ",")
		data = append(data, Context{maps, nums})
	}
	cache = make(map[Cache]int)

	fmt.Println("Part 1:", part1(data))
	fmt.Println("Part 2:", part2(data))
}
